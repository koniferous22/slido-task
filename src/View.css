body {
	background: rgb(245,245,245);
	color: #1e1309;
}

div#container {
	display: flex;
	flex-direction: column;

	font-family: Verdana, sans-serif;
}


/*
* case list
*/

section.cases {
    grid-area: cases;
    text-align: center;
    
    display: flex;
    justify-content: center;
    align-items: center;
}

section.cases div#cases-hint {
	flex-shrink: 2;
	margin: auto 0;

}

@media screen and (max-width: 600px) {
	section.cases div#cases-hint {
		display: none;
	}
}

section.cases section {
	/*margin: auto 4vw;*/
	max-width: 30vw;
	overflow-x: auto;
	flex-grow: 30000;
}

section.cases section#past {
	transform: matrix3d(
    	1, 0, 0, 0.001,
    	0.1, 1, 0, 0,
    	0, 0, 1, 0,
    	0, 0, 0, 1);
}

section.cases section#future {
	transform: matrix3d(
    	1, 0, 0, -0.001,
    	-0.1, 1, 0, 0,
    	0, 0, 1, 0,
    	0, 0, 0, 1);
	/*margin-right: 50px;*/
}

section.cases ul.case-list {
    max-height: 20vh;
	overflow-y: scroll;
    padding: 10px 5vw;
    border-radius: 20px;

    transition: transform .2s;
    /*disable scrollbar in firefox*/
    scrollbar-width: none;

    list-style-type: none;
}

section.cases ul.case-list:hover {
	scrollbar-width: thin;
}

/*disable scrollbar in webkit based browsers*/
section.cases ul.case-list::-webkit-scrollbar {
   display: none;
}


section.cases ul.case-list:after {
    content: "";
    display: block;
    width: 100%;
    height: 100%;
    background: linear-gradient(to bottom, rgba(245,245,245,0) 0%, rgba(245,245,245,0) 70%, rgba(245,245,245,1) 100%);
    pointer-events: none;
    position: absolute;
    top: 0;
    left: 0;
}

section.cases ul.case-list li {
	transition: transform .2s;
}

section.cases ul.case-list li:hover {
	transform: scale(1.1);
	cursor: pointer;
}

section#content {
	grid-area: content;
	text-align: center;
	margin-top: 50px;
	max-width: 80vw;
	margin: auto;
	/*fixes overlaying fade from case list*/
	z-index: 1;
}

form#case-add {
	grid-area: form;

	display: grid;
	grid-template-columns: 100px auto;

	text-align: center;
	margin: auto;
	width: 66vw;
	max-width: 600px;
}


form#case-add h3#form-title {
	grid-column: 1/3;
	text-align: center;
}

form#case-add div.error {
	grid-column: 1/3;
	text-align: center;
	font-weight: bold;
	color: red;
}

form#case-add input.submit-button {
	text-align: center;
	background-color: transparent;
	grid-column: 1/3;
	
	margin: 20px auto 10px auto;
	border-radius: 3px;
	border: 2px solid #51300a;

	font-weight: bold;
}


form#case-add input {
	border-radius: 5px;
}

form#case-add textarea {
	border-radius: 5px;
	resize: none;
}



form#case-add input.submit-button:hover {
	text-decoration: underline;
	cursor: pointer;
}
