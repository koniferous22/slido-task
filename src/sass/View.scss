@use 'base';

body {
	background: base.$background-color;
}

div#container {
	display: flex;
	flex-direction: column;
	font-family: base.$font;
	color: base.$font-color;
	text-align: center;
}
