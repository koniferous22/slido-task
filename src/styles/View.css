body {
	background: rgb(245,245,245);
	color: #1e1309;
}

div#container {
	display: flex;
	flex-direction: column;

	font-family: Verdana, sans-serif;
}

div#container section#cases {
	display: flex;
	justify-content: center;
	align-items: center;
	text-align: center;
}

section#content {
	text-align: center;
	max-width: 80vw;
	word-wrap: break-word;
	overflow-wrap: break-word;
	margin: 50px auto auto auto;
	/*fixes overlaying fade from case list*/
	z-index: 1;
}

form#case-add {
	grid-area: form;

	display: grid;
	grid-template-columns: 100px auto;

	text-align: center;
	margin: auto;
	width: 66vw;
	max-width: 600px;
}


form#case-add h3#form-title {
	grid-column: 1/3;
	text-align: center;
}

form#case-add div.error {
	grid-column: 1/3;
	text-align: center;
	font-weight: bold;
	color: red;
}

form#case-add input.submit-button {
	text-align: center;
	background-color: transparent;
	grid-column: 1/3;
	
	margin: 20px auto 10px auto;
	border-radius: 3px;
	border: 2px solid #51300a;

	font-weight: bold;
}


form#case-add input {
	border-radius: 5px;
}

form#case-add textarea {
	border-radius: 5px;
	resize: none;
}

form#case-add input.submit-button:hover {
	text-decoration: underline;
	cursor: pointer;
}
