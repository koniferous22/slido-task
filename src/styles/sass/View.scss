@use 'base';

body {
	background: base.$background-color;
}

div#container {
	display: flex;
	flex-direction: column;
	font-family: base.$font;
	color: base.$font-color;
	text-align: center;
}
div#container section#cases {
	display: flex;
	justify-content: center;
	align-items: center;
}
div#container section#content {
	@include base.page-section(50px, 80vw);
	@include base.fix-text-wrap;
	/*fixes overlaying fade from case list*/
	z-index: 1;
}
