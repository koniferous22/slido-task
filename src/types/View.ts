import { Case } from './common/Case'

export interface ViewState {
	cases: Case[];
	selectedCaseIndex: number;
}
