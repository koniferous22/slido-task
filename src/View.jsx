import React from 'react'

import CaseList from './components/CaseList'
import FormInput from './components/FormInput'
import FormTextarea from './components/FormTextarea'

import './View.css'

import { sortedArrayItemInsertionIndex } from './helper'

class View extends React.Component {
	constructor() {
		super()
		this.defaultFormValues = {
			title: '',
			description: '',
			date: ''
		}
		this.state = {
			// we assume this array is sorted by date
			cases: [],
			selectedCase: undefined,
			// decided not to use nested state, because of the overhead when accessing attributes
			formValues: { ...this.defaultFormValues },
			// for now empty, so that the messages are undefined
			formErrorFlags: { }
		}

		this.handleSubmitEvent = this.handleSubmitEvent.bind(this)
		this.selectDisplayedCase = this.selectDisplayedCase.bind(this)
		this.handleFormChangeEvent = this.handleFormChangeEvent.bind(this)
		this.handleFormBlurEvent = this.handleFormBlurEvent.bind(this)
	}

	validateFormFields(...fields) {
		const reduceCb = (flagz, field) => (typeof field !== 'string' && this.state.formValues[field]) ? flagz : Object.assign(flagz, {[field]: this.state.formValues[field] === ''})
		const newFormErrorFlags = fields.reduce(
			reduceCb, this.state.formErrorFlags
		)
		console.log(newFormErrorFlags)
		this.setState({
			formErrorFlags: newFormErrorFlags
		})
	}
	
	handleFormChangeEvent(formField, event) {
		this.setState({
			formValues: Object.assign(this.state.formValues, {[formField]: event.target.value})
		})
	}

	handleFormBlurEvent(formField) {
		this.validateFormFields(formField);
	}

	handleSubmitEvent(event) {
		// prevents page reloading on form submit
		event.preventDefault();
		
		this.validateFormFields('title', 'description', 'date')
		const parsers = {
			title: value => value,
			description: value => value,
			date: value => new Date(value)
		}

		// verify that every error has been fixed
		if (Object.keys(this.state.formErrorFlags).every(key => this.state.formErrorFlags[key] === false)) {
			const newCase = Object.fromEntries(
				Object.entries(this.state.formValues).map(([key, value]) => [key, parsers[key](value)])
			)

			let insertionIndex = sortedArrayItemInsertionIndex(this.state.cases, newCase, (c1, c2) => c1.date < c2.date)
			// inserts in to the array
			this.state.cases.splice(insertionIndex, 0, newCase);

			this.setState({
				// keeps the order of the selected item
				selectedCase: this.state.selectedCase === undefined ? 0 : ( insertionIndex <= this.state.selectedCase ? this.state.selectedCase + 1 : this.state.selectedCase ),
				// reset the form here, also forces rerender
				formValues: { ...this.defaultFormValues },
				formErrorFlags: {}
			})
		}
	}

	selectDisplayedCase(selectedCaseIndex) {
		if (selectedCaseIndex < this.state.cases.length && selectedCaseIndex >= 0) {
			this.setState({
				selectedCaseIndex
			})
		}
	}

	render() {
		const displayPast = this.state.selectedCase > 0
		const displayFuture = this.state.selectedCase + 1 < this.state.cases.length
		return (
			<div id="container">
				<section className="cases">
					{
						displayPast && <CaseList id="past" caseEntries={this.state.cases.slice(0, this.state.selectedCaseIndex)} handleSelectCase={this.selectDisplayedCase}/>
					}
					{
						this.state.cases.length > 1 &&
							<div id="cases-hint">
								<h4>Pick an event to display</h4>
								<h6>lists are scrollable :)</h6>
							</div>
					}
					{
						displayFuture && 
							<CaseList 
								id="future" 
								caseEntries={this.state.cases.slice(this.state.selectedCaseIndex + 1)} 
								handleSelectCase={this.selectDisplayedCase} 
								offset={this.state.selectedCaseIndex + 1}
							/>
					}
				</section>
				<section id="content">
				{
					// conition whether array index is in range already verified in event hadler "handle select case"
					// first condition does not catch when state.selectedCase is equal to zero
					(this.state.selectedCase || this.state.selectedCase === 0) ? (
							<>
								<h2 id="content-title">{this.state.cases[this.state.selectedCase].title}</h2>
								<i>{this.state.cases[this.state.selectedCase].date.toLocaleDateString()}</i>
								<p>{'Description: ' + this.state.cases[this.state.selectedCase].description}</p>
							</>
						) : (
							<>
								Submit form down below to add events
							</>
						)
				}
				</section>
				<form id="case-add" onSubmit={this.handleSubmitEvent}>
					<h3 id="form-title">Submit event here</h3>
					<FormInput
						name="title" label="Title"
						type="text"
						errorFlag={this.state.formErrorFlags.title}
						onBlur={this.handleFormBlurEvent}
						onChange={this.handleFormChangeEvent}
						value={this.state.formValues.title || ''}
					/>
					<FormTextarea
						name="description" label="Description"
						errorFlag={this.state.formErrorFlags.description}
						onBlur={this.handleFormBlurEvent}
						onChange={this.handleFormChangeEvent}
						rows={5}
						value={this.state.formValues.description || ''}
					/>
					<FormInput
						name="date" label="Event Date"
						type="date"
						errorFlag={this.state.formErrorFlags.date}
						onBlur={this.handleFormBlurEvent}
						onChange={this.handleFormChangeEvent}
						value={this.state.formValues.date || ''}
					/>
					<input className="submit-button" type="submit" value="Add Event" />
				</form>
			</div>
		)
	}
}

export default View
