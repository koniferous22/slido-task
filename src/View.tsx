import React from 'react'

import { ViewState } from './types/ViewState'
import { Case } from './types/Case'
import { Dictionary } from './types/Dictionary'

import CaseList from './components/CaseList'
import FormInput from './components/FormInput'
import FormTextArea from './components/FormTextArea'

import './View.css'

import { sortedArrayItemInsertionIndex } from './helper'


const defaultFormValues: Dictionary<string> = {
	title: '',
	description: '',
	date: ''
}

class View extends React.Component<{}, ViewState> {
	constructor(props: {}) {
		super(props)
		this.state = {
			// we assume this array is sorted by date
			cases: [],
			selectedCaseIndex: 0,
			// decided not to use nested state, because of the overhead when accessing attributes
			formValues: { ...defaultFormValues },
			// for now empty, so that the messages are undefined
			formErrorFlags: { }
		}

		this.handleSubmitEvent = this.handleSubmitEvent.bind(this)
		this.selectDisplayedCase = this.selectDisplayedCase.bind(this)
		this.handleFormChangeEvent = this.handleFormChangeEvent.bind(this)
		this.handleFormBlurEvent = this.handleFormBlurEvent.bind(this)
	}

	validateFormFields(...fields: string[]) {
		const reduceCb = (flagz: Dictionary<boolean>, field: string): Dictionary<boolean> => {
			return /*const result = this.state.formValues[field] ? flagz :*/ Object.assign(flagz, {[field]: (this.state.formValues[field] === '')})
		}
		const newFormErrorFlags = fields.reduce(
			reduceCb, this.state.formErrorFlags
		)
		this.setState({
			formErrorFlags: newFormErrorFlags
		})
	}
	
	handleFormChangeEvent(formField: string, event: React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>) {
		this.setState({
			formValues: Object.assign(this.state.formValues, {[formField]: event.target.value.trim()})
		})
	}

	handleFormBlurEvent(formField: string) {
		this.validateFormFields(formField)
	}

	handleSubmitEvent(event: React.FormEvent<HTMLFormElement>) {
		// prevents page reloading on form submit
		event.preventDefault();
		
		this.validateFormFields('title', 'description', 'date')

		// verify that every error has been fixed
		if (Object.keys(this.state.formErrorFlags).every((key: string) => this.state.formErrorFlags[key] === false)) {
			// Previous revision contains cooler way to do this, but had to resolve typescript and stuff like that
			// still probably better, cause lower overhead (not worth applying arrow functions on 3 items :D )
			const newCase: Case = {
				title: this.state.formValues.title,
				description: this.state.formValues.description,
				date: new Date(this.state.formValues.date)
			}
			const insertionIndex: number = sortedArrayItemInsertionIndex(this.state.cases, newCase, (c1: Case, c2: Case): boolean => c1.date < c2.date)
			// inserts in to the array
			this.state.cases.splice(insertionIndex, 0, newCase)

			this.setState({
				// keeps the order of the selected item
				selectedCaseIndex:  insertionIndex <= this.state.selectedCaseIndex ? this.state.selectedCaseIndex + 1 : this.state.selectedCaseIndex,
				// reset the form here, also forces rerender
				formValues: { ...defaultFormValues },
				formErrorFlags: {}
			})
		}
	}

	selectDisplayedCase(selectedCaseIndex: number) {
		if (selectedCaseIndex < this.state.cases.length && selectedCaseIndex >= 0) {
			this.setState({
				selectedCaseIndex
			})
		}
	}

	render() {
		const displayPast: boolean = this.state.selectedCaseIndex > 0
		const displayFuture: boolean = this.state.selectedCaseIndex + 1 < this.state.cases.length
		return (
			<div id="container">
				<section className="cases">
					{
						displayPast && <CaseList id="past" caseEntries={this.state.cases.slice(0, this.state.selectedCaseIndex)} handleSelectCase={this.selectDisplayedCase}/>
					}
					{
						this.state.cases.length > 1 &&
							<div id="cases-hint">
								<h4>Pick an event to display</h4>
								<h6>lists are scrollable :)</h6>
							</div>
					}
					{
						displayFuture && 
							<CaseList 
								id="future" 
								caseEntries={this.state.cases.slice(this.state.selectedCaseIndex + 1)} 
								handleSelectCase={this.selectDisplayedCase} 
								offset={this.state.selectedCaseIndex + 1}
							/>
					}
				</section>
				<section id="content">
				{
					this.state.cases[this.state.selectedCaseIndex] ? (
							<>
								<h2 id="content-title">{this.state.cases[this.state.selectedCaseIndex].title}</h2>
								<i>{this.state.cases[this.state.selectedCaseIndex].date.toLocaleDateString()}</i>
								<p>{'Description: ' + this.state.cases[this.state.selectedCaseIndex].description}</p>
							</>
						) : (
							<>
								Submit form down below to add events
							</>
						)
				}
				</section>
				<form id="case-add" onSubmit={this.handleSubmitEvent}>
					<h3 id="form-title">Submit event here</h3>
					<FormInput
						name="title" label="Title"
						type="text"
						errorFlag={this.state.formErrorFlags.title}
						onBlur={this.handleFormBlurEvent}
						onChange={this.handleFormChangeEvent}
						value={this.state.formValues.title || ''}
					/>
					<FormTextArea
						name="description" label="Description"
						errorFlag={this.state.formErrorFlags.description}
						onBlur={this.handleFormBlurEvent}
						onChange={this.handleFormChangeEvent}
						rows={5}
						value={this.state.formValues.description || ''}
					/>
					<FormInput
						name="date" label="Event Date"
						type="date"
						errorFlag={this.state.formErrorFlags.date}
						onBlur={this.handleFormBlurEvent}
						onChange={this.handleFormChangeEvent}
						value={this.state.formValues.date || ''}
					/>
					<input className="submit-button" type="submit" value="Add Event" />
				</form>
			</div>
		)
	}
}

export default View
