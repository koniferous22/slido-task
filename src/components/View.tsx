import React from 'react'

import { ViewState } from 'types/View'
import { Case } from 'types/common/Case'

import Cases from './Cases'
import Content from './Content'
import CaseAddForm from './CaseAddForm'

import 'styles/View.css'

import { getSortedInsertionIndex } from 'utils/functions'

class View extends React.Component<{}, ViewState> {
	constructor(props: {}) {
		super(props)

		this.state = {
			// we assume this array is sorted by date
			cases: [],
			selectedCaseIndex: 0,
		}

		this.addCase = this.addCase.bind(this)
		this.selectDisplayedCase = this.selectDisplayedCase.bind(this)
	}

	addCase(case_: Case): void {
		const {cases, selectedCaseIndex} = this.state

		const insertionIndex: number = getSortedInsertionIndex(cases, case_, (c1: Case, c2: Case): boolean => c1.date < c2.date)
		// inserts in to the array
		cases.splice(insertionIndex, 0, case_)

		const shouldSelectedIndexShift: boolean = insertionIndex <= selectedCaseIndex && insertionIndex + 1 < cases.length
		const newSelectedCase: number = shouldSelectedIndexShift ? selectedCaseIndex + 1 : selectedCaseIndex
		this.setState({
			selectedCaseIndex: newSelectedCase,
		})
	}

	selectDisplayedCase(selectedCaseIndex: number): void {
		if (selectedCaseIndex < this.state.cases.length && selectedCaseIndex >= 0) {
			this.setState({selectedCaseIndex})
		}
	}

	render(): React.ReactNode {
		const {cases, selectedCaseIndex} = this.state
		return (
			<div id="container">
				<Cases
					caseEntries={cases}
					selectedCaseIndex={selectedCaseIndex}
					handleSelectCase={this.selectDisplayedCase}
				/>
				<Content
					placeholder="Submit form down below to add events"
					displayedCase={cases[selectedCaseIndex]}
				/>
				<CaseAddForm onSubmit={this.addCase}/>
			</div>
		)
	}
}

export default View
