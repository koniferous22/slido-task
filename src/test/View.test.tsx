import React from 'react'
import renderer from 'react-test-renderer'
import View from 'components/View'

it('View renders correctly', () => {	
	const tree = renderer.create(<View/>).toJSON();
	expect(tree).toMatchSnapshot();
});