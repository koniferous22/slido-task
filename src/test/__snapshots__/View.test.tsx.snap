// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View renders correctly 1`] = `
<div
  id="container"
>
  <section
    id="cases"
  />
  <section
    id="content"
  >
    Submit form down below to add events
  </section>
  <form
    id="case-add"
    onSubmit={[Function]}
  >
    <h3
      id="form-title"
    >
      Submit event here
    </h3>
    <label
      htmlFor="title"
    >
      Title: 
    </label>
    <input
      id="title"
      name="title"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value=""
    />
    <label
      htmlFor="description"
    >
      Description: 
    </label>
    <textarea
      id="description"
      name="description"
      onBlur={[Function]}
      onChange={[Function]}
      rows={5}
      value=""
    />
    <label
      htmlFor="date"
    >
      Event Date: 
    </label>
    <input
      id="date"
      name="date"
      onBlur={[Function]}
      onChange={[Function]}
      type="date"
      value=""
    />
    <input
      className="submit-button"
      type="submit"
      value="Add Event"
    />
  </form>
</div>
`;
